spinalcordtoolbox/manual-correction
# Generalize manual correction scripts

## Description

This PR proposes an initial version of scripts for manual correction workflow. For details, see [README](https://github.com/spinalcordtoolbox/manual-correction/blob/jv/generalize_manual_correction_scripts/README.md).
The scripts are based on [ukbiobank-spinalcord-csa](https://github.com/sct-pipeline/ukbiobank-spinalcord-csa) scripts. To allow easy tracking of changes proposed in this PR, I pushed the `ukbiobank-spinalcord-csa` scripts to the [main branch](https://github.com/spinalcordtoolbox/manual-correction).

Compared `ukbiobank-spinalcord-csa` scripts, I added:
- new input flags (`-path-derivatives`,  `-path-derivatives`, `-suffix-files-seg`, `-suffix-files-gmseg`, `-suffix-files-label`, `-label-list`, `-viewer`) in https://github.com/spinalcordtoolbox/manual-correction/commit/f89a9aceca97c7dc6aedf35e3aa950eb8da7ad91
- correction of gmseg in https://github.com/spinalcordtoolbox/manual-correction/commit/291be5f2257af344e1066934698037181be923ad
- selection of viewer (FSLeyes, ITK-SNAP. 3D Slicer is not implemented yet, though) in https://github.com/spinalcordtoolbox/manual-correction/commit/44a54f6937600365f545335cb033ca80bdf881cc
- `fetch_subject_and_session` function (to unify `get_subject` and `get_session` under a single function) in https://github.com/spinalcordtoolbox/manual-correction/commit/0e312aa402f19c4dda09aafc2118c406eae5b84c and https://github.com/spinalcordtoolbox/manual-correction/commit/daf74dcd7ed889f84b326533ff5fdae3df79757c
- `fetch_yaml_config` function to simplify loading of yaml config file among scripts in https://github.com/spinalcordtoolbox/manual-correction/commit/272c4c9cba622172e70edfd027a578a821ea1cfa
- further improvements (such as manipulation with paths, comments to clarify code, etc.)

I would extremely appreciate feedback from SCT users, such as @sandrinebedard, @kiristern, @naga-karthik, @rohanbanerjee, and @mchen1110. You can test the PR using the instructions below or on some dataset you are currently working on.

This PR partly addresses #1. But still, a lot of features describes in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3353#issuecomment-1380916712 have to be implemented. 

## How to test this PR

1. Clone repo, create `venv`, and install dependencies:

```
git clone https://github.com/spinalcordtoolbox/manual-correction.git
cd manual-correction
git checkout jv/generalize_manual_correction_scripts
python3 -m venv venv
source venv/bin/activate
pip install -r requirements.txt
```

2. Download the [testing dataset](https://www.dropbox.com/s/adkgjqgxn2wj4iq/manual_correction_example_dataset.zip?dl=0). The dataset contains a sample output as provided by `sct_run_batch`. The dataset also includes the `files_to_correct.yml` file with a list of files for manual correction (obtained from the [SCT QC html report](https://spinalcordtoolbox.com/overview/concepts/inspecting-results-qc-fsleyes.html#how-do-i-use-the-qc-report).

3. Run `package_for_correction.py`:

This simulates the real-world workflow in which a dataset is processed on a remote server, and you do not want to copy the whole dataset (GB of data). Instead, only the images and labels that need to be corrected (listed in `files_to_correct.yml`) are zipped by the `package_for_correction.py`:

```
python package_for_correction.py -path-in ~/manual_correction_example_dataset/data_processed -config ~/manual_correction_example_dataset/files_to_correct.yml -o ~/manual_correction_example_dataset/data_to_correct
```

(Now, you would have copied the zipped file from a server to your local machine. Since we are just simulating real-world workflow, the copy from a server to a local machine is skipped for now.)

4. Run `manual_correction.py`:

The `package_for_correction.py` outputs `data_to_correct.zip`. Unzip this archive and run `manual_correction.py`:

```
python manual_correction.py -path-in ~/manual_correction_example_dataset/data_to_correct -config ~/manual_correction_example_dataset/files_to_correct.yml -path-out ~/Desktop/manual_correction_example_dataset -viewer fsleyes
```

Note: you can specify a viewer (FSLeyes, ITK-SNAP) for SC and GM segmentation corrections. (3D Slicer is not implemented yet)

First, you will have to enter your name (which will be used for .json sidecars). Then, the `package_for_correction.py` will iterate across files listed in `files_to_correct.yml`. It will gradually open FSLeyes (or ITK-snap) for SC and GM segmentation correction and sct_label_utils for disc label correction. The script saves manually corrected files under the `derivatives/labels` folder (you can change this folder by the `-path-derivatives` flag). The script also outputs a QC report (--> you can check how you corrected the labels).

Once finished, you can rerun the command. In this case, the script detects manually corrected files under `derivatives/labels` and asks you if you want to overwrite them.

Fixes: https://github.com/spinalcordtoolbox/manual-correction/issues/1
